src/Model/Table/SentencesTable.php

```diff
@@ -1105,19 +1127,6 @@ public function changeLanguage($sentenceId, $newLang)
         if (($ownerId == $currentUserId || CurrentUser::isModerator()) && !$this->hasAudio($sentence->id)) {
             $sentence->lang = $newLang;
             $result = $this->save($sentence);
-
-            $this->Contributions->updateLanguage($sentenceId, $newLang);
```

Comment:
Correct me if I'm wrong, but this line suggests that `Contributions->updateLanguage()` is never called when the language is edited from the new design (which uses `editSentence()` instead) and that your PR fixes this problem. So we might have to fix languages in the contributions table just like in #2387. Could you confirm this and add a unit test?

Reply:
Yes, `editSentence` didn't do all the changes that `changeLanguage` did.
A few weeks ago I've added [sentence #7841656 on dev](https://dev.tatoeba.org/deu/sentences/show/7841656) as Latin:
```
MariaDB [tatoeba]> select lang from sentences where id = 7841656;
+------+
| lang |
+------+
| lat  |
+------+
1 row in set (0.001 sec)

MariaDB [tatoeba]> select * from contributions where sentence_id = 7841656;
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
| sentence_id | sentence_lang | translation_id | translation_lang | script | text         | action | user_id | datetime            | type     | id       |
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
|     7841656 | lat           |           NULL | NULL             | NULL   | lorem ipsum  | insert |   70604 | 2020-04-24 14:20:04 | sentence | 26641857 |
|     7841656 | NULL          |           NULL | NULL             | NULL   | CC BY 2.0 FR | insert |   70604 | 2020-04-24 14:20:04 | license  | 26641858 |
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
2 rows in set (0.001 sec)
```
Now I've changed the language to German using the new design:
```
MariaDB [tatoeba]> select lang from sentences where id = 7841656;
+------+
| lang |
+------+
| deu  |
+------+
1 row in set (0.001 sec)

MariaDB [tatoeba]> select * from contributions where sentence_id = 7841656;
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
| sentence_id | sentence_lang | translation_id | translation_lang | script | text         | action | user_id | datetime            | type     | id       |
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
|     7841656 | lat           |           NULL | NULL             | NULL   | lorem ipsum  | insert |   70604 | 2020-04-24 14:20:04 | sentence | 26641857 |
|     7841656 | NULL          |           NULL | NULL             | NULL   | CC BY 2.0 FR | insert |   70604 | 2020-04-24 14:20:04 | license  | 26641858 |
+-------------+---------------+----------------+------------------+--------+--------------+--------+---------+---------------------+----------+----------+
2 rows in set (0.000 sec)
```

But now that you have asked I wonder if the old behavior (changing the languages in `contributions` for old rows) is really a good idea. I've considered the entries in `contributions` as read-only and the log would allow us to reconstruct how a sentence changed over time. But "rewriting the history" makes this impossible.

Shouldn't we rather implement #533?

Furthermore even using the old design we do not update the translation language in `contributions` ([7842015](https://dev.tatoeba.org/deu/sentences/show/7842015), [7842016](https://dev.tatoeba.org/deu/sentences/show/7842016)):
Before:
```
MariaDB [tatoeba]> select id, lang, created, modified from sentences where id in (7842015, 7842016);
+---------+------+---------------------+---------------------+
| id      | lang | created             | modified            |
+---------+------+---------------------+---------------------+
| 7842015 | deu  | 2020-07-13 04:57:39 | 2020-07-13 04:57:39 |
| 7842016 | eng  | 2020-07-13 04:58:01 | 2020-07-13 04:58:01 |
+---------+------+---------------------+---------------------+
2 rows in set (0.001 sec)

MariaDB [tatoeba]> select sentence_id, sentence_lang, translation_id, translation_lang, action, type, datetime from contributions where sentence_id in (7842015
, 7842016);
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
| sentence_id | sentence_lang | translation_id | translation_lang | action | type     | datetime            |
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
|     7842015 | deu           |           NULL | NULL             | insert | sentence | 2020-07-13 04:57:39 |
|     7842015 | NULL          |           NULL | NULL             | insert | license  | 2020-07-13 04:57:40 |
|     7842015 | NULL          |        7842016 | NULL             | insert | link     | 2020-07-13 04:58:01 |
|     7842016 | eng           |           NULL | NULL             | insert | sentence | 2020-07-13 04:58:01 |
|     7842016 | NULL          |           NULL | NULL             | insert | license  | 2020-07-13 04:58:01 |
|     7842016 | NULL          |        7842015 | NULL             | insert | link     | 2020-07-13 04:58:01 |
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
6 rows in set (0.001 sec)
```
After:
```
MariaDB [tatoeba]> select id, lang, created, modified from sentences where id in (7842015, 7842016);
+---------+------+---------------------+---------------------+
| id      | lang | created             | modified            |
+---------+------+---------------------+---------------------+
| 7842015 | rus  | 2020-07-13 04:57:39 | 2020-07-13 05:05:18 |
| 7842016 | spa  | 2020-07-13 04:58:01 | 2020-07-13 05:05:22 |
+---------+------+---------------------+---------------------+
2 rows in set (0.001 sec)

MariaDB [tatoeba]> select sentence_id, sentence_lang, translation_id, translation_lang, action, type, datetime from contributions where sentence_id in (7842015, 7842016);
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
| sentence_id | sentence_lang | translation_id | translation_lang | action | type     | datetime            |
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
|     7842015 | rus           |           NULL | NULL             | insert | sentence | 2020-07-13 04:57:39 |
|     7842015 | rus           |           NULL | NULL             | insert | license  | 2020-07-13 04:57:40 |
|     7842015 | rus           |        7842016 | NULL             | insert | link     | 2020-07-13 04:58:01 |
|     7842016 | spa           |           NULL | NULL             | insert | sentence | 2020-07-13 04:58:01 |
|     7842016 | spa           |           NULL | NULL             | insert | license  | 2020-07-13 04:58:01 |
|     7842016 | spa           |        7842015 | NULL             | insert | link     | 2020-07-13 04:58:01 |
+-------------+---------------+----------------+------------------+--------+----------+---------------------+
6 rows in set (0.001 sec)
```
On dev the most recent entry in `contributions` that has a `translation_lang` is from 2010:
```
MariaDB [tatoeba]> select max(datetime) from contributions where length(translation_lang) > 0;
+---------------------+
| max(datetime)       |
+---------------------+
| 2010-04-06 19:28:07 |
+---------------------+
1 row in set (14.789 sec)
```

And you can also see that calling `Contributions->updateLanguage()` modifies `sentence_lang` for every row whereas we don't store a value for `sentence_lang` when we log license insertion and link insertion.

I'm afraid contributions logging needs some fixing. So do you still want to have the unit test in this PR?

Reply:
About `contributions.translation_lang`, maybe this field was used somehow in the past, but now we are not displaying this information anywhere, so it makes sense that it is currently not updated. I think we could as well remove that column.

As for `contributions.sentence_lang`, you have a point. It’s okay not to include a unit test.